### PR TITLE
feat(upunk): add skeleton of uPUNK price feed

### PR DESCRIPTION
**Pull request by pemulis**
Signed-off-by: John Shutt <john.d.shutt@gmail.com>

**Motivation**

This adds the skeleton of an expiry price feed for the uPUNK index, defined as `PUNKETH` in UMIP-84 (https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-84.md).

**Summary**

This strips down the `UniswapPriceFeed` and grabs on-chain event data in a similar way. Remaining work: finish the internal function to compute the median price from the event array, taking only the most recent price for each unique punk. See UMIP-84 for details.

**Details**

Pre-expiry price requests (`PUNKETH_TWAP`) will simply use the `UniswapPriceFeed` to get the self-referential price for the uPUNK synthetic token. A financial product library will fetch the price using the `PUNKETH` identifier after expiry.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [x]  Untested

**Issue(s)**

Fixes #XXXX

### CODE REVIEW

File: packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
+// An implementation of PriceFeedInterface that calculates the expiry price of the uPUNK index over a given period.
+
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { averageBlockTimeSeconds } = require("@uma/common");
+class UpunkPriceFeed extends PriceFeedInterface {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js:5]: Should this be named "UPunk"? (note capitalization)

Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js:5]: Or maybe even UPUNK

File: packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    // If there are still no prices, return null to allow the user to handle the absence of data.
+    if (events.length === 0) {
+      this.events = [];
+      return;
+    }
+
+    // Filter out events where price is null.
+    this.events = events.filter(e => e.price !== null);
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js:132]: Maybe you should filter out null prices before the `events.length === 0` check, so we don't try to call `this._computePrice` on an empty array

**Review by nicholaspai — APPROVED** [review]
So to be clear, we're going to use this price feed in the `optimistic-oracle-proposer` bot to propose an expiry price? And this is preferable to a more manual calculation using one of chase's python scripts right? 

The implementation looks very clean though and its very easy to read. Curious how you're going to implement `computePrice` :)

File: packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      }
+
+      return a.logIndex - b.logIndex;
+    });
+
+    return events;
+  }
+
+  _computePrice(/* eventsIn, startTime, endTime */) {
+    // const events = eventsIn.map(e => {
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js:159]: is this meant to be commented out?

**Comment by chrismaree** [comment]
Any updates on this PR? it's been sitting idle for a long time @pemulis

**Comment by pemulis** [comment]
> Any updates on this PR? it's been sitting idle for a long time @pemulis

The expiry price calculation needs to be filled in, but it's dropped lower in priority than some other DevX items. We have expiry calculators in Go and Python already and have some time to set up the JS version before expiry.

**Comment by chrismaree** [comment]
> > Any updates on this PR? it's been sitting idle for a long time @pemulis
> 
> The expiry price calculation needs to be filled in, but it's dropped lower in priority than some other DevX items. We have expiry calculators in Go and Python already and have some time to set up the JS version before expiry.

ok cool. ideally, it's not ideal to have PRs sitting in open and requested review state for weeks on end. perhaps consider closing the PR and re-opening it when you are ready for full review or trying to get it closed? Not great to clog the open PRs in the repo.

**Comment by pemulis** [comment]
Closing for now. Will re-open later when I'm working on it again.

**Pull request closed without merging**